content/blog/2024-08-01-unikraft-gsoc-debugging-mimalloc.mdx

```diff
@@ -0,0 +1,132 @@
+
+Unikraft's current virtual memory interface is not without its flaws.
+I ran into multiple "Cannot handle read page fault" errors during boot time.
+
+```console
+[   11.290098] Info: <init> {r:0x175768,f:0x11e70} test: posix_futex_testsuite->test_wait_timeout
+[   11.304300] dbg:  <init> {r:0x130da1,f:0x11ba0} [libposix_futex] <futex.c @  283> (int) uk_syscall_r_futex((uint32_t *) 0x11e48, (int) 0x0, (uint32_t) 0xa, (const struct timespec *) 0x11e30, (uint32_t *) 0x0, (uint32_t) 0x0)
+[   11.326513] dbg:  <init> {r:0x13094b,f:0x11af0} [libposix_futex] <futex.c @  112> FUTEX_WAIT: Condition met (*uaddr == 10, uaddr: 0x11e48)
+[   11.342579] dbg:  <init> {r:0x1309ae,f:0x11af0} [libposix_futex] <futex.c @  124> FUTEX_WAIT: Wait 12326512446 nsec for wake-up
+[   11.358184] dbg:  <idle> {r:0x207f6f,f:0x119d0} [libcontext] <sysctx.c @   37> Sysctx 0x1000025fc8 GS_BASE register updated to 0x29e140 (before: 0x11bf0)
+[   11.375007] CRIT: <idle> {r:0x17dde9,f:0x28ce70} [libukvmem] <pagefault.c @   43> Cannot handle exec page fault at 0x0 (ec: 0x10): Bad address (14).
+[   11.392395] CRIT: <idle> {r:0x106f17,f:0x28cf00} [libkvmplat] <trace.c @   41> RIP: 0000000000000000 CS: 0008
+[   11.407999] CRIT: <idle> {r:0x106f6c,f:0x28cf00} [libkvmplat] <trace.c @   42> RSP: 000000100003afe8 SS: 0010 EFLAGS: 00010246
+[   11.424430] CRIT: <idle> {r:0x106fb8,f:0x28cf00} [libkvmplat] <trace.c @   44> RAX: 0000000000000000 RBX: 0000000000000000 RCX: 0000000000000000
+[   11.441611] CRIT: <idle> {r:0x107004,f:0x28cf00} [libkvmplat] <trace.c @   46> RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
+[   11.458821] CRIT: <idle> {r:0x107050,f:0x28cf00} [libkvmplat] <trace.c @   48> RBP: 0000000000011a70 R08: 0000000000000000 R09: 0000000000000000
+[   11.476011] CRIT: <idle> {r:0x10709c,f:0x28cf00} [libkvmplat] <trace.c @   50> R10: 0000000000000000 R11: 000000000000002d R12: 0000000000000000
+[   11.493210] CRIT: <idle> {r:0x1070e8,f:0x28cf00} [libkvmplat] <trace.c @   52> R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
+[   11.510388] CRIT: <idle> {r:0x107240,f:0x28cf00} [libkvmplat] <trace.c @   86> base is 0x11a70 caller is 0x1732e6
+```
+
+After extensive debugging, I realized that this bug is not unique to the *mimalloc* port.
+I reproduced a similar error with the same library configurations on the buddy allocator as well and I have submitted [an issue](https://github.com/unikraft/unikraft/issues/1475) addressing this.
+More work needs to be done to fix the virtual memory infrastructure.
+
+## Next steps
+
+In order to resolve the abovementioned issues, I still have much to study about Unikraft's virtual memory management subsystem, its boot process, and how *mimalloc* works under the hood - and I will focus on just that in the following weeks.
+
+Hopefully, with those issues fixed, we could finally get the `cache-scratch` test running on multiple cores.
+Once that is done, I will get back on track with my "global" TODO list:
```

Comment:
Can you mention what ideas do you have for benchmarking?

Reply:
I have written much about benchmarking (i.e., which benchmarks I would use, what they are for, etc.) in my last blog post so I didn't elaborate here. I will attach a link to my last blog post